Drivers/Src/CANbus.c

```diff
@@ -1,61 +1,201 @@
+       case VELOCITY:
+       case MC_PHASE_CURRENT:
+       case VOLTAGE_VEC:
+       case CURRENT_VEC:
+       case BACKEMF:
+       case TEMPERATURE:
+       case ODOMETER_AMPHOURS:
+            datalen = 8;
+            memcpy(&txdata,&payload.data.d,sizeof(payload.data.d));
 
-    int8_t data[payload.bytes];
+    //Handle 8bit precision case (0b0000xxxx) (no idx)
+       case CAR_STATE:
+            datalen = 1;
+            memcpy(&txdata,&payload.data.b,sizeof(payload.data.b)); //copy into txdata array  the 8 bits of the payload
 
-    uint64_t tempData = payload.data.d;
-    uint8_t mask = 0xFF;
+    }
 
-    for(int i=payload.bytes-1; i>=0; i--){
-        data[i] = tempData&mask;
-        tempData = tempData>>8;
+    OSMutexPend( //ensure that tx line is available
+        &CANbus_TxMutex,
+        0,
+        OS_OPT_PEND_BLOCKING,
+        &timestamp,
+        &err
+    );
+    if (err != OS_ERR_NONE){
+        //couldn't lock tx line
+        return ERROR;
     }
+    //tx line locked
+
+    int retval = BSP_CAN_Write(CAN_1, id, txdata, datalen);
 
+    OSMutexPost( //unlock the TX line
+        &CANbus_TxMutex,
+        OS_OPT_POST_NONE,
+        &err
+    );
 
-	return BSP_CAN_Write(CAN_1, id, data, payload.bytes);
+    if (retval){
+        return SUCCESS;
+    } else {
+        return ERROR;
+    }
 }
 
 
 /**
  * @brief   Checks if the CAN ID matches with expected ID and then copies message to given buffer array
- * @param   CAN line bus
+ * @param   CAN line bus (variable buses not implemented, this function currently works with CAN1)
+ * @param   ID pointer to where to store the CAN id of the recieved msg
  * @param   pointer to buffer array to store message
+ * @param   blocking whether or not this read should be blocking
  * @return  1 if ID matches and 0 if it doesn't
  */
 
-ErrorStatus CANbus_Read(uint8_t* buffer){
-    
-    uint32_t ID;
-    uint8_t data[8];
-    uint8_t count = BSP_CAN_Read(CAN_1,&ID,data);
+ErrorStatus CANbus_Read(CAN_t canLine,CANId_t *ID,uint8_t* buffer,CAN_blocking_t blocking){
+    CPU_TS timestamp;
+    OS_ERR err;
+
+    if(blocking == CANBLOCKING){
+        OSSemPend(
+            &CANBus_RecieveSem4,
+            0,
+            OS_OPT_PEND_BLOCKING,
+            &timestamp,
+            &err
+        );
+    } else {
+        OSSemPend(
+            &CANBus_RecieveSem4,
+            0,
+            OS_OPT_PEND_NON_BLOCKING,
+            &timestamp,
+            &err
+        );
+    }
+    if(err != OS_ERR_NONE) {
+        return ERROR;
+    }
+
+    OSMutexPend( //ensure that RX line is available
+        &CANbus_RxMutex,
+        0,
+        OS_OPT_PEND_BLOCKING,
+        &timestamp,
+        &err
+    );
+    if (err != OS_ERR_NONE){ 
+        //couldn't lock RX line
+        return ERROR;
+    }
+
+    //GETMSG
+    uint8_t status = BSP_CAN_Read(CAN_1,ID,buffer);
     
-    if(ID == MOTOR_DISABLE){
-        for(int i=0;i<count;i++){
-            buffer[i]=data[i];
-        }

```

Comment:
This function should still only return a `SUCCESS` if the message that was read was the correct ID. The open PR #140 has changed what specific ID this is, but it's still expecting this function to return based on the ID. I would wait for that PR to merge in then resolve any conflicts and keep the ID check.